contrib/firebaseio/src/main/java/contrib/JacksonCoder.java

```diff
@@ -0,0 +1,153 @@
+import com.fasterxml.jackson.databind.ObjectReader;
+import com.fasterxml.jackson.databind.ObjectWriter;
+import com.fasterxml.jackson.databind.SerializationFeature;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Encodes objects using {@link ObjectReader} and {@link ObjectWriter}.
+ */
+public class JacksonCoder<T> extends StandardCoder<T> {
+
+  private static final long serialVersionUID = -754345287170755870L;
+
+  private static final ObjectMapper MAPPER = new ObjectMapper()
+      .enableDefaultTyping(DefaultTyping.NON_FINAL)
+      .enable(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS)
+      .enable(DeserializationFeature.WRAP_EXCEPTIONS)
+      .enable(SerializationFeature.WRAP_EXCEPTIONS)
+      .disable(SerializationFeature.FAIL_ON_UNWRAPPED_TYPE_IDENTIFIERS)
+      .enable(SerializationFeature.WRITE_NULL_MAP_VALUES)
+      .configure(JsonGenerator.Feature.AUTO_CLOSE_TARGET, false)
+      .configure(JsonParser.Feature.AUTO_CLOSE_SOURCE, false);
+
+  // Threadsafe as per http://wiki.fasterxml.com/JacksonBestPracticesPerformance
+  private final ObjectWriter writer;
+  private final ObjectReader reader;
+
+  protected final Class<T> type;
+
+  public static <K> JacksonCoder<K> of(TypeDescriptor<K> type) {
+    @SuppressWarnings("unchecked")
+    Class<K> clazz = (Class<K>) type.getRawType();
+    return of(clazz);
+  }
+
+  /**
+   * JacksonCoder will encode any type that can be serialized/deserialized by Jackson.
+   * This can be achieved by using either Jackson annotations (as seen in this library), or
+   * implementing @see <a href=http://wiki.fasterxml.com/JacksonHowToCustomSerializers>
+   * custom serializers</a>.
+   * @param clazz Type to encode
+   * @return A JacksonCoder parameterized with type {@code clazz}
+   */
+  public static <K> JacksonCoder<K> of(Class<K> clazz){
+    return new JacksonCoder<>(clazz);
+  }
+
+  @SuppressWarnings({"unchecked", "rawtypes"})
+  @JsonCreator
+  public static JacksonCoder<?> of(
+      @JsonProperty("type") String classType) throws ClassNotFoundException {
+    return new JacksonCoder(Class.forName(classType));
+  }
+
+  protected JacksonCoder(Class<T> clazz){
+    this.type = clazz;
+    writer = MAPPER.writer();
+    reader = MAPPER.readerFor(type);
+  }
+
+  @Override
+  public void encode(T value, OutputStream outStream,
+      com.google.cloud.dataflow.sdk.coders.Coder.Context context)
+          throws IOException {
+    writer.writeValue(outStream, value);
+  }
+
+  @Override
+  public T decode(InputStream inStream, com.google.cloud.dataflow.sdk.coders.Coder.Context context)
+      throws IOException {
+    return reader.readValue(inStream);
+  }
+
+  @Override
+  public List<? extends Coder<?>> getCoderArguments() {
+    return Collections.emptyList();
+  }
+
+  @Override
+  public void verifyDeterministic() {  }
```

Comment:
This should throw, as JSON is not a deterministic encoding (even with sorted keys).

Is it important that this coder be deterministic? Determinism is used in `GroupByKey`, but you should not be using complex JSON-encodings for the keys in a GBK. You should use small, concisely-encoded values for performance.